### DISCUSSION TITLE
[Q] How client(React.js) download excel file in NodeJS?

### DISCUSSION THREAD

**Discussion by resobyte** [exceljs/exceljs#2396] [Q&A]
## 💬 Questions and Help

Hi everyone,

I have two separate projects. Client: React, Server: NodeJS

I create excel on the NodeJS side by submitting a form by React.

I want to download this excel from React side. However, I couldn't download it.

### NodeJS code

```
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader("Content-Disposition", "attachment; filename=" + "Report.xlsx");
  workbook.xlsx.write(res)
      .then(function(){
          res.end();
      });
```
### NodeJS return network layer
<img width="472" alt="Screen Shot 2020-12-07 at 19 39 28" src="https://user-images.githubusercontent.com/17431441/101378977-c97ba080-38c4-11eb-96af-8e458fad3e5c.png">

### First try React.js code

```
 startDownload(response,"resobyte.xlsx")

function startDownload(file, fileName) {
    const url = window.URL.createObjectURL(new Blob([file]));
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', fileName);
    document.body.appendChild(link);
    link.click();
    link.parentNode.removeChild(link);
}
```

### Second try React.js code

```
  let blob = new Blob([response], {type: 'vnd.openxmlformats-officedocument.spreadsheetml.sheet;charset=utf-8'});
            FileSaver.saveAs(blob, 'fileName.xlsx");
```

The file has been downloaded.

### My React Service Call

```
export const createExcel = async (model) => {
  let response = () => {
      return new Promise(function(resolve, reject) {
        fetch(API_URL + '/api/excel/create', {
          method: 'POST',
          responseType: 'arrayBuffer',
          headers: {
          'Content-Type': 'application/json',
          'x-access-token' : TokenService.getToken()
          },
          body: JSON.stringify(model),
        }).then(response => {
          resolve(response);
        });
      });
    };
    let responseData = await response();
    return responseData;
}

```
### Error open excel file

<img width="421" alt="Screen Shot 2020-12-07 at 19 49 55" src="https://user-images.githubusercontent.com/17431441/101379566-77874a80-38c5-11eb-9978-39344398d37c.png">

**Please do not originate issue 37. Because I tried all of them don't work.**

**Comment by aksharj** [comment]
@resobyte, did you fix this yet, i am facing a similar issue, my environment is node js

**Comment by resobyte** [comment]
Hi @aksharj,

**React.js**

```
export const getExcel = async (id,isSendMail = true) => {
  axios.get(API_URL + '/api/excel/getExcel/'+id+'?isSendMail='+isSendMail, { responseType: 'arraybuffer' , headers: {
    'x-access-token' : TokenService.getToken()
    }})
    .then((response) => {
      const dirtyFileName = response.headers['content-disposition'];
      const regex = /filename[^;=\n]*=(?:(\\?['"])(.*?)\1|(?:[^\s]+'.*?')?([^;\n]*))/;
      var  fileName = dirtyFileName.match(regex)[3];
    
      var blob = new Blob([response.data], { type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' });
      fileSaver.saveAs(blob, fileName);
    });
}
```

**Node.js**

```
   res.setHeader('Access-Control-Expose-Headers', "Content-Disposition"); //IMPORTANT FOR React.js content-disposition get Name
   res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
   res.setHeader("Content-Disposition", "attachment; filename=" + fileNameExcel+".xlsx");
    return workbook.xlsx.write(res)
        .then(function(){
            res.end();
        });
      })
```

You can fix this code.

I used axios and fileSaver. You can use **GET** method.

Important: responseType :'arrayBuffer'

**Comment by aksharj** [comment]
hi @resobyte, thanks for sharing the solution. Does this work with large files, if i understand correctly, you are loading the entire contents of the file in memory.

**Comment by resobyte** [comment]
Hi @aksharj,

I didn't try large files. My excel file 30KB.

If you try, can you give me information too..

Thanks.
Best regards

**Comment by TkachenkoDmitry** [comment]
@resobyte
Please try this one:

```ts
if(linkRef.current){
    const linkSource = `data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,${data.productsReport}`;
    const downloadLink = linkRef.current;
    const fileName = `${reportName}.xlsx`;
    downloadLink.download= fileName;
    downloadLink.href = linkSource;
    downloadLink.target = '_blank';
    downloadLink.click();
}
```

Create an `a` tag and use its a ref to create needed properties. 

on Back End 

```ts
const buffer = await workbook.xlsx.writeBuffer() as unknown as Buffer;

return buffer.toString('base64');
```

Note that casting to unknown is needed because the `Buffer` is not defined correctly.

Don't hesitate to reach me if you have any questions

**Comment by yepMad** [comment]
After some tests I came to a conclusion.

On my backend:
```ts
const buffer = (await workbook.xlsx.writeBuffer()) as Buffer;
response.write(buffer);
response.end();
```

On my front-end:

```ts
axios.get(`myRoute`, { responseType: 'arraybuffer' })
      .then(response => {
        const blob = new Blob([response.data], {
          type:
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        });

        FileDownload(blob, `file.xlsx`); // npm " js-file-download" responsible for downloading the file for the user
      });
```

**Comment by mudassiriqball** [comment]
You can do this by fetch API.
I am using Next.js so I have placed my .xslx file into the public folder.

    const downloadBulkUploadTemplete = async () => {
        fetch("SpreadsheeName.xlsx", {
            method: 'GET',
        }).then(response => response.blob())
            .then(blob => {
                var url = window.URL.createObjectURL(blob);
                var a = document.createElement('a');
                a.href = url;
                a.download = "bac.xlsx";
                document.body.appendChild(a); // we need to append the element to the dom -> otherwise it will not work in firefox
                a.click();
                a.remove();  //afterwards we remove the element again 
            });
        }

**Comment by lpcuong2106** [comment]
> After some tests I came to a conclusion.
> 
> On my backend:
> 
> ```ts
> const buffer = (await workbook.xlsx.writeBuffer()) as Buffer;
> response.write(buffer);
> response.end();
> ```
> 
> On my front-end:
> 
> ```ts
> axios.get(`myRoute`, { responseType: 'arraybuffer' })
>       .then(response => {
>         const blob = new Blob([response.data], {
>           type:
>             'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
>         });
> 
>         FileDownload(blob, `file.xlsx`); // npm " js-file-download" responsible for downloading the file for the user
>       });
> ```

FileDownload Not found function

**Comment by yepMad** [comment]
> > After some tests I came to a conclusion.
> > On my backend:
> > ```ts
> > const buffer = (await workbook.xlsx.writeBuffer()) as Buffer;
> > response.write(buffer);
> > response.end();
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On my front-end:
> > ```ts
> > axios.get(`myRoute`, { responseType: 'arraybuffer' })
> >       .then(response => {
> >         const blob = new Blob([response.data], {
> >           type:
> >             'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
> >         });
> > 
> >         FileDownload(blob, `file.xlsx`); // npm " js-file-download" responsible for downloading the file for the user
> >       });
> > ```
> 
> FileDownload Not found function

Are you sure you are using "import FileDownload from 'js-file-download';"?

**Comment by ghost** [comment]
> After some tests I came to a conclusion.
> 
> On my backend:
> 
> ```ts
> const buffer = (await workbook.xlsx.writeBuffer()) as Buffer;
> response.write(buffer);
> response.end();
> ```
> 
> On my front-end:
> 
> ```ts
> axios.get(`myRoute`, { responseType: 'arraybuffer' })
>       .then(response => {
>         const blob = new Blob([response.data], {
>           type:
>             'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
>         });
> 
>         FileDownload(blob, `file.xlsx`); // npm " js-file-download" responsible for downloading the file for the user
>       });
> ```

my result is like that:
![image](https://user-images.githubusercontent.com/13423042/154725469-9918f9f8-e1b7-4bbf-bb6a-773d80290198.png)

what am I missing?

**Comment by yepMad** [comment]
> my result is like that: ![image](https://user-images.githubusercontent.com/13423042/154725469-9918f9f8-e1b7-4bbf-bb6a-773d80290198.png)
> 
> what am I missing?

This looks like a buffer to me. Show me your code.

**Comment by ghost** [comment]
> > my result is like that: ![image](https://user-images.githubusercontent.com/13423042/154725469-9918f9f8-e1b7-4bbf-bb6a-773d80290198.png)
> > what am I missing?
> 
> This looks like a buffer to me. Show me your code.

Thanks for the answer.
Here is the code

Node:
```
const dataToExport = res.message.result.data;
                const workbook = new ExcelJS.Workbook();
                workbook.created = new Date();
                workbook.views = [
                    {
                        x: 0, y: 0, width: 10000, height: 20000,
                        firstSheet: 0, activeTab: 1, visibility: 'visible'
                    }
                ];
                const sheet = workbook.addWorksheet('data1');
                sheet.columns = [];
                function capitalizeFirstLetter(string) {
                    return string.charAt(0).toUpperCase() + string.slice(1);
                }
                const columns = [];
                for (const key in dataToExport[0]) {
                    columns.push({ header: capitalizeFirstLetter(key), key: key});
                }

                sheet.columns = columns;
                dataToExport.forEach((el)=> {
                    for (const key in el) {
                        if (el[key] === null) {
                            el[key] = '';
                        }
                        if (typeof el[key] === 'boolean') {
                            el[key] = el[key].toString();
                        }
                    }

                });
                sheet.addRows(dataToExport);
                workbook._worksheets.forEach(ws => { // makes column larger
                    if (ws._columns) {
                        ws._columns.forEach(column => {
                            let maxWidth = 5;
                            column.values.forEach(value => {
                                maxWidth = value.length > maxWidth ? value.length : maxWidth;
                            });
                            column.width = maxWidth + 3;
                        });
                    }
                });
                try {
                    workbook.xlsx.writeBuffer()
                        .then((buffer) => {
                            const oJson = {};
                            oJson['result'] = buffer;
                            resolve({code: 200, message: oJson});

                        });
                }
```

Frontend:
```
.then((response) => {
        const blob = new Blob([response.data], {
          type:
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;charset=UTF-8'
        })
        fileDownload(blob, 'file.xlsx')
```

**Comment by himarnoel** [comment]
.then((response) => {
        const blob = new Blob([response.data], {
          type:
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;charset=UTF-8'
        })
        fileDownload(blob, 'file.xlsx')

I used it but the file does not open

**Comment by NazirAnsari** [comment]
Any one fo this for post method??
